attestation-service/verifier/src/eventlog/mod.rs

```diff
@@ -0,0 +1,186 @@
+        let init_state = hex::decode(init_state).context("parse init state in INIT entry")?;
+
+        let events = eventline
+            .iter()
+            .map(|line| AAEvent::from_str(line))
+            .collect::<Result<Vec<AAEvent>>>()?;
+
+        Ok(Self {
+            events,
+            hash_algorithm,
+            init_state,
+        })
+    }
+}
+
+impl AAEventlog {
+    fn accumulate_hash<D: Digest + FixedOutput>(&self) -> Vec<u8> {
+        let mut state = self.init_state.clone();
+
+        let mut init_event_hasher = D::new();
+        let init_event = format!(
+            "INIT {}/{}",
+            self.hash_algorithm.as_ref(),
+            hex::encode(&self.init_state)
+        );
+        Digest::update(&mut init_event_hasher, init_event.as_bytes());
+        let init_event_hash = init_event_hasher.finalize();
+
+        let mut hasher = D::new();
+        Digest::update(&mut hasher, &state);
+
+        Digest::update(&mut hasher, init_event_hash);
+        state = hasher.finalize().to_vec();
+
+        self.events.iter().for_each(|event| {
+            let mut event_hasher = D::new();
+            Digest::update(&mut event_hasher, event.domain.as_bytes());
+            Digest::update(&mut event_hasher, b" ");
+            Digest::update(&mut event_hasher, event.operation.as_bytes());
+            Digest::update(&mut event_hasher, b" ");
+            Digest::update(&mut event_hasher, event.content.as_bytes());
+            let event_hash = event_hasher.finalize();
+
+            let mut hasher = D::new();
+            Digest::update(&mut hasher, &state);
+            Digest::update(&mut hasher, event_hash);
+            state = hasher.finalize().to_vec();
+        });
+
+        state
+    }
+
+    /// Check the integrity of the AAEL, and gets a digest. The digest should be the same
+    /// as the input `rtmr`, or the integrity check will fail.
+    pub fn integrity_check(&self, rtmr: &[u8]) -> Result<()> {
+        let result = match self.hash_algorithm {
+            HashAlgorithm::Sha256 => self.accumulate_hash::<Sha256>(),
+            HashAlgorithm::Sha384 => self.accumulate_hash::<Sha384>(),
+            HashAlgorithm::Sha512 => self.accumulate_hash::<Sha512>(),
+        };
+
+        if rtmr != result {
+            bail!(
+                "AA eventlog does not pass check. AAEL value : {}, Quote value {}",
+                hex::encode(result),
+                hex::encode(rtmr)
+            );
+        }
+
+        Ok(())
+    }
+
+    pub fn to_parsed_claims(&self) -> Map<String, Value> {
+        let mut aael = Map::new();
+        for eventlog in &self.events {
+            let key = format!("{}/{}", eventlog.domain, eventlog.operation);
+            let item = Value::String(eventlog.content.clone());
+            match aael.get_mut(&key) {
+                Some(value) => value
+                    .as_array_mut()
+                    .expect("Only array can be inserted")
```

Comment:
should we panic here?

Reply:
This is some logic that will never panic. The initial `aael` is empty. If a key is `None`, a `Value::Array` will be inserted. Thus whenever a `Some` is found, it must be a `Value::Array`.

Here, the same `operation` could have multiple different events, thus we make all different events an `Array` of a same `domain/operation`.

Reply:
mmh I understand that it would be an internal (programmer) error that should never happen. future refactorings might modify the logic and introduce breakage. and then it's often nicer to look at an error log line rather than a crashed AS (at least that's my experience with the tdx attester). But yeah, that's arguable whether we want to panic on internal errors, I don't think we follow a common practice here, so i'm fine with either.

Reply:
You are right that risks will hide here. Let me add a comment to tell the future refactoring writers.